rpc/rest/security/rest_security.go

```diff
@@ -0,0 +1,415 @@
+	// Returns login page where credentials may be put. Redirects to authenticate, and if successful, moves to index.
+	login = "/login"
+	// URL key for logout, invalidates current token.
+	logout = "/logout"
+	// Authentication page, validates credentials and if successful, returns a token or writes a cookie to a browser
+	authenticate = "/authenticate"
+	// Cookie name identifier
+	cookieName = "ligato-rest"
+)
+
+// Default value to sign the token, if not provided from config file
+var signature = "secret"
+
+// Default expiration time for token/cookie
+var expTime time.Duration = 3600000000000 // 1 Hour
+
+// AuthenticatorAPI provides methods for handling permissions
+type AuthenticatorAPI interface {
+	// AddPermissionGroup adds new permission group. PG is defined by name and a set of URL keys. User with
+	// permission group enabled has access to that set of keys. PGs with duplicated names are skipped.
+	AddPermissionGroup(group ...*httpsecurity.PermissionGroup)
+
+	// Validate serves as middleware used while registering new HTTP handler. For every request, token
+	// and permission group is validated.
+	Validate(provider http.HandlerFunc) http.HandlerFunc
+}
+
+// Context defines fields required to instantiate authenticator
+type Context struct {
+	StorageType StorageType
+	Users       []httpsecurity.User
+	ExpTime     time.Duration
+	Cost        int
+	Signature   string
+}
+
+// Credentials struct represents simple user login input
+type credentials struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+}
+
+// Authenticator keeps information about users, permission groups and tokens and processes it
+type authenticator struct {
+	log logging.Logger
+
+	// Router instance automatically registers login/logout REST API handlers if authentication is enabled
+	router    *mux.Router
+	formatter *render.Render
+
+	// User database keeps all known users with permissions and hashed password. Users are loaded from
+	// HTTP config file
+	// TODO add option to register users
+	userDb AuthStore
+	// Permission database is a map of name/permissions and bound URLs
+	groupDb map[string][]*httpsecurity.PermissionGroup_Permissions
+
+	// Token claims
+	expTime time.Duration
+}
+
+// NewAuthenticator prepares new instance of authenticator.
+func NewAuthenticator(router *mux.Router, ctx *Context, log logging.Logger) AuthenticatorAPI {
+	a := &authenticator{
+		router: router,
+		log:    log,
+		formatter: render.New(render.Options{
+			IndentJSON: true,
+		}),
+		userDb:  CreateAuthStore(ctx.StorageType),
+		groupDb: make(map[string][]*httpsecurity.PermissionGroup_Permissions),
+		expTime: ctx.ExpTime,
+	}
+
+	// Set token signature
+	signature = ctx.Signature
+	if a.expTime == 0 {
+		a.expTime = expTime
+		a.log.Debugf("Token expiration time claim not set, defaulting to 1 hour")
+	}
+
+	// Add admin-user, enabled by default, always has access to every URL
+	hash, err := bcrypt.GenerateFromPassword([]byte("ligato123"), ctx.Cost)
```

Comment:
This might potentially be expensive operation. Could you do it in goroutine? but keep in mind that needs to be concurrently safe, perhaps at the auth store level.